modules/utxo-lib/test/bitgo/psbt/psbtUtil.ts

```diff
@@ -0,0 +1,278 @@
+        );
+      } catch (err) {
+        if (err.message === 'unsigned p2tr sign should fail') {
+          throw err;
+        }
+        assert.deepStrictEqual(signatureTarget, 'unsigned');
+        assert.deepStrictEqual(scriptTypeForChain(u.chain), 'p2tr');
+        assert.deepStrictEqual(psbt.data.inputs[i].tapLeafScript, undefined);
+        assert.deepStrictEqual(psbt.data.inputs[i].tapBip32Derivation, undefined);
+        assert.deepStrictEqual(psbt.data.inputs[i].tapScriptSig, undefined);
+        assert.ok(psbt.data.inputs[i].witnessUtxo);
+      }
+    } else {
+      throw new Error(`invalid unspent`);
+    }
+  });
+}
+
+export function signTxBuilder<TNumber extends number | bigint>(
+  txb: UtxoTransactionBuilder<TNumber, UtxoTransaction<TNumber>>,
+  unspents: Unspent<TNumber>[],
+  rootWalletKeys: RootWalletKeys,
+  signer: string,
+  cosigner: string,
+  signatureTarget: SignatureTargetType
+): UtxoTransaction<TNumber> {
+  let walletUnspentSigners: WalletUnspentSigner<RootWalletKeys>[] = [];
+
+  if (signatureTarget === 'halfsigned') {
+    walletUnspentSigners = [WalletUnspentSigner.from(rootWalletKeys, rootWalletKeys[signer], rootWalletKeys[cosigner])];
+  } else if (signatureTarget === 'fullsigned') {
+    walletUnspentSigners = [
+      WalletUnspentSigner.from(rootWalletKeys, rootWalletKeys[signer], rootWalletKeys[cosigner]),
+      WalletUnspentSigner.from(rootWalletKeys, rootWalletKeys[cosigner], rootWalletKeys[signer]),
+    ];
+  }
+
+  walletUnspentSigners.forEach((walletSigner, nSignature) => {
+    unspents.forEach((u, i) => {
+      if (isWalletUnspent(u)) {
+        signInputWithUnspent(txb, i, u, walletSigner);
+      } else {
+        throw new Error(`unexpected unspent ${u.id}`);
+      }
+    });
+  });
+
+  return signatureTarget === 'fullsigned' ? txb.build() : txb.buildIncomplete();
+}
+
+export function constructTransactionUsingTxBuilder<TNumber extends number | bigint>(
+  unspents: Unspent<TNumber>[],
+  rootWalletKeys: RootWalletKeys,
+  signer: string,
+  cosigner: string,
+  amountType: 'number' | 'bigint' = 'number',
+  outputType: outputScripts.ScriptType2Of3,
+  signatureTarget: SignatureTargetType,
+  params: {
+    network: Network;
+    changeIndex: number;
+    fee: bigint;
+  }
+): UtxoTransaction<bigint> {
+  const txb = createTransactionBuilderForNetwork<TNumber>(params.network);
+  const total = BigInt(unspentSum<TNumber>(unspents, amountType));
+  // Kinda weird, treating entire value as change, but tests the relevant paths
+  txb.addOutput(
+    getWalletAddress(rootWalletKeys, getInternalChainCode(outputType), params.changeIndex, params.network),
+    toTNumber<TNumber>(total - params.fee, amountType)
+  );
+  unspents.forEach((u) => {
+    addToTransactionBuilder(txb, u);
+  });
+
+  return signTxBuilder<TNumber>(txb, unspents, rootWalletKeys, signer, cosigner, signatureTarget).clone<bigint>(
+    'bigint'
+  );
+}
+
+export function mockUnspents<TNumber extends number | bigint>(
```

Comment:
Looks generic enough that it could live outside the `psbt` scope. Should it live in `test/bitgo/wallet/util.ts`?

Reply:
Done